tests/test_generics.py

```diff
@@ -1174,9 +1185,9 @@ class Model(BaseModel, Generic[T, KT, VT]):
 
     # The following assertions are just to document the current behavior, and should
     # be updated if/when we do a better job of respecting the exact annotated type
-    assert type(m.counter_field) is Counter.__origin__
+    assert type(m.counter_field) is CustomCounter
     assert type(m.default_dict_field) is dict
-    assert type(m.deque_field) is deque
+    assert type(m.deque_field) is CustomDeque
```

Comment:
I think we have several of these inconsistencies where we handled subclasses in some cases but not others. The main impetus for this PR is to resolve these things by not doing the wrong thing silently and instead requiring some input from users on how they want us to handle their custom types.
Given how much easier it is now to handle both first-party and third-party custom types in V2 compared to V1, this is a very reasonable thing to do.

(this was all discussed offline, I'm just leaving it here for the record)

Reply:
I think the most important thing given this change is making sure users have an easy time finding the docs for custom types and/or examples like this (even if just in tests)

Reply:
We actually point you to `__get_pydantic_core_schema__` in the error you get, but I agree an explicit test that shows that would be good in this PR